File: npmjs/pray-cli/src/manifest/deprecation.ts
export const DEPRECATED_TARGET = "target";
export const DEPRECATED_OUTPUT = "output";
export const DEPRECATED_AGENT = "agent";
export const DEPRECATED_SKILLS = "skills";
export const DEPRECATED_SKILL = "skill";
export const DEPRECATED_SPEC_SKILLS = "spec.skills";

const REPLACEMENTS: Record<string, string> = {
  [DEPRECATED_TARGET]: "compose` / `tree",
  [DEPRECATED_OUTPUT]: "compose",
  [DEPRECATED_AGENT]: "pray",
  [DEPRECATED_SKILLS]: "tree` / `folder",
  [DEPRECATED_SKILL]: "folder",
  [DEPRECATED_SPEC_SKILLS]: "a folder export",
};

export function noteDeprecatedKeyword(
  keywords: string[] | undefined,
  keyword: string,
): string[] {
  if (!(keyword in REPLACEMENTS)) {
    return keywords ?? [];
  }
  const current = keywords ?? [];
  if (current.includes(keyword)) {
    return current;
  }
  return [...current, keyword];
}

export function deprecationWarnings(keywords: string[] | undefined): string[] {
  return (keywords ?? [])
    .map((keyword) => {
      const replacement = REPLACEMENTS[keyword];
      if (!replacement) {
        return undefined;
      }
      return `warning: \`${keyword}\` is deprecated and will be removed in version 2; prefer \`${replacement}\``;
    })
    .filter((warning): warning is string => warning !== undefined);
}

export function emitDeprecationWarnings(keywords: string[] | undefined): void {
  for (const warning of deprecationWarnings(keywords)) {
    console.error(warning);
  }
}

export function emitResolvedDeprecationWarnings(
  keywords: string[] | undefined,
  packages: Array<{
    spec: {
      skills: Map<string, unknown>;
      exports: Map<string, { kind: string }>;
    };
  }>,
): void {
  emitDeprecationWarnings(keywords);
  const seen = new Set<string>();
  for (const resolved of packages) {
    for (const warning of packageSpecDeprecationWarnings(resolved.spec)) {
      if (seen.has(warning)) {
        continue;
      }
      seen.add(warning);
      console.error(warning);
    }
  }
}

export function packageSpecDeprecationWarnings(spec: {
  skills: Map<string, unknown>;
  exports: Map<string, { kind: string }>;
}): string[] {
  const keywords: string[] = [];
  if (spec.skills.size > 0) {
    keywords.push(DEPRECATED_SPEC_SKILLS);
  }
  for (const exportEntry of spec.exports.values()) {
    if (exportEntry.kind === DEPRECATED_SKILL) {
      keywords.push(DEPRECATED_SKILL);
      break;
    }
  }
  return deprecationWarnings(keywords);
}

File: crates/pray-core/src/deprecation.rs
use crate::package_spec::PackageSpec;

/// Prayfile keywords retained for compatibility and scheduled for removal in version 2.
pub const DEPRECATED_TARGET: &str = "target";
pub const DEPRECATED_OUTPUT: &str = "output";
pub const DEPRECATED_AGENT: &str = "agent";
pub const DEPRECATED_SKILLS: &str = "skills";
pub const DEPRECATED_SKILL: &str = "skill";
pub const DEPRECATED_SPEC_SKILLS: &str = "spec.skills";

pub fn is_deprecated_prayfile_keyword(keyword: &str) -> bool {
    matches!(
        keyword,
        DEPRECATED_TARGET | DEPRECATED_OUTPUT | DEPRECATED_AGENT | DEPRECATED_SKILLS
    )
}

pub fn deprecation_warning(keyword: &str) -> Option<String> {
    let replacement = match keyword {
        DEPRECATED_TARGET => "compose` / `tree",
        DEPRECATED_OUTPUT => "compose",
        DEPRECATED_AGENT => "pray",
        DEPRECATED_SKILLS => "tree` / `folder",
        DEPRECATED_SKILL => "folder",
        DEPRECATED_SPEC_SKILLS => "a folder export",
        _ => return None,
    };
    Some(format!(
        "warning: `{keyword}` is deprecated and will be removed in version 2; prefer `{replacement}`"
    ))
}

pub fn package_spec_deprecation_warnings(spec: &PackageSpec) -> Vec<String> {
    let mut keywords = Vec::new();
    if !spec.skills.is_empty() {
        keywords.push(DEPRECATED_SPEC_SKILLS.to_string());
    }
    if spec
        .exports
        .values()
        .any(|export| export.kind == DEPRECATED_SKILL)
    {
        keywords.push(DEPRECATED_SKILL.to_string());
    }
    deprecation_warnings_for(&keywords)
}

pub fn deprecation_warnings_for(keywords: &[String]) -> Vec<String> {
    let mut warnings = Vec::new();
    for keyword in keywords {
        if let Some(warning) = deprecation_warning(keyword) {
            warnings.push(warning);
        }
    }
    warnings
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn warns_for_legacy_keywords() {
        let warnings = deprecation_warnings_for(&[
            DEPRECATED_TARGET.to_string(),
            DEPRECATED_OUTPUT.to_string(),
            DEPRECATED_AGENT.to_string(),
            DEPRECATED_SKILLS.to_string(),
        ]);
        assert_eq!(warnings.len(), 4);
        assert!(warnings[0].contains("`target`"));
        assert!(warnings[0].contains("version 2"));
        assert!(warnings[1].contains("`output`"));
        assert!(warnings[2].contains("`agent`"));
        assert!(warnings[2].contains("`pray`"));
        assert!(warnings[3].contains("`skills`"));
        assert!(warnings[3].contains("`tree`"));
    }
}
